### PR TITLE
Feature/125 welcome letter

### PR DESCRIPTION
**Pull request by SashaBondarchuk**
image

https://trello.com/c/Oy08mJN1/125-welcome-letter

### CODE REVIEW

File: backend/LeetWars.Emailer/Interfaces/IMailService.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+﻿using LeetWars.Core.Common.DTO.Mail;
+
+namespace LeetWars.Emailer.Interfaces
+{
+    public interface IMailService
+    {
+        Task<IResult> SendWithValidationCheck(MailDto mailDto);
```
Comment (Limbo2332) [inline backend/LeetWars.Emailer/Interfaces/IMailService.cs]:
```suggestion
        Task<IResult> SendWithValidationCheckAsync(MailDto mailDto);
```

File: backend/LeetWars.Emailer/Services/MailService.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+        private async Task SendEmailAsync(MailDto mailDto)
+        {
+            var senderEmail = new EmailAddress(_configuration["Email"], _configuration["EmailSenderName"]);
+            var receiverEmail = new EmailAddress(mailDto.ReceiverEmail);
+            var message = MailHelper.CreateSingleEmail(senderEmail, receiverEmail, mailDto.Subject, mailDto.Content, mailDto.Content);
+            await _sendGridClient.SendEmailAsync(message);
+        }
+
+        public async Task<IResult> SendWithValidationCheck(MailDto mailDto)
```
Comment (Limbo2332) [inline backend/LeetWars.Emailer/Services/MailService.cs]: see previous comments

File: frontend/src/app/core/services/email.service.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import { Injectable } from '@angular/core';
+import { Mail } from '@shared/models/email/email';
+
+import { HttpInternalService } from './http-internal.service';
+
+@Injectable({
+    providedIn: 'root',
+})
+export class EmailService {
+    public baseUrl: string = 'http://localhost:5080/emailer';
```
Comment (Limbo2332) [inline frontend/src/app/core/services/email.service.ts]: Move localhost:5080 to environment file and reuse it here

**Review by Limbo2332 — CHANGES_REQUESTED** [review]

File: backend/LeetWars.Core/LeetWars.Core.BLL/Services/HttpService.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
+﻿using LeetWars.Core.BLL.Interfaces;
+using System.Text;
+
+namespace LeetWars.Core.BLL.Services
+{
+    public class HttpService : IHttpService
+    {
+        private readonly HttpClient _httpClient;
```
Comment (Limbo2332) [inline backend/LeetWars.Core/LeetWars.Core.BLL/Services/HttpService.cs]: Dispose httpclient

**Review by Limbo2332 — CHANGES_REQUESTED** [review]

**Review by Limbo2332 — APPROVED** [review]

File: backend/LeetWars.Core/LeetWars.Core.BLL/Services/HttpService.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
+﻿using LeetWars.Core.BLL.Interfaces;
+using System.Text;
+
+namespace LeetWars.Core.BLL.Services
+{
+    public class HttpService : IHttpService
```
Comment (tatianahutii) [inline backend/LeetWars.Core/LeetWars.Core.BLL/Services/HttpService.cs]: why do we need HTTP client if we have RabbitMQ for communication between these two services?

File: backend/LeetWars.Core/LeetWars.Core.WebAPI/Controllers/EmailerController.cs [hunk 6]
```diff
@@ -0,0 +1,30 @@
+﻿using LeetWars.Core.BLL.Interfaces;
+using LeetWars.Core.Common.DTO.Mail;
+using Microsoft.AspNetCore.Mvc;
+using Newtonsoft.Json;
+
+namespace LeetWars.Core.WebAPI.Controllers
+{
+    [Route("[controller]")]
+    [ApiController]
+    public class EmailerController : ControllerBase
```
Comment (tatianahutii) [inline backend/LeetWars.Core/LeetWars.Core.WebAPI/Controllers/EmailerController.cs]: are you sure we need this comtroller?

File: backend/LeetWars.Emailer/Dockerfile [hunk 7]
```diff
@@ -1,13 +1,15 @@
-FROM mcr.microsoft.com/dotnet/aspnet:6.0 AS base
+FROM mcr.microsoft.com/dotnet/aspnet:6.0-focal AS base
```
Comment (tatianahutii) [inline backend/LeetWars.Emailer/Dockerfile:1]: why did you change this?

File: backend/LeetWars.Core/LeetWars.Core.WebAPI/appsettings.json [hunk 8]
```diff
@@ -14,6 +14,7 @@
   },
   "AllowedHosts": "*",
   "Rabbit": "amqp://guest:guest@localhost:5672",
+  "EmailApiURI": "http://localhost:5080/emailer/sendEmail",
```
Comment (tatianahutii) [inline backend/LeetWars.Core/LeetWars.Core.WebAPI/appsettings.json]: it would be better to save host separately

File: backend/LeetWars.Emailer/Program.cs [hunk 9]
```diff
@@ -1,6 +1,27 @@
+using FluentValidation;
+using LeetWars.Core.Common.DTO.Mail;
+using LeetWars.Emailer.Interfaces;
+using LeetWars.Emailer.Services;
+using LeetWars.Emailer.Validators;
+
 var builder = WebApplication.CreateBuilder(args);
+
+builder.Services.AddScoped<IMailService, MailService>();
+builder.Services.AddScoped<IValidator<MailDto>, MailDtoValidator>();
```
Comment (tatianahutii) [inline backend/LeetWars.Emailer/Program.cs]: to the separate extension method

File: frontend/src/app/core/services/email.service.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+@Injectable({
+    providedIn: 'root',
+})
+export class EmailService {
+    public baseUrl: string = '/emailer';
+
+    constructor(private httpService: HttpInternalService) { }
+
+    public sendEmail(mail: Mail) {
```
Comment (tatianahutii) [inline frontend/src/app/core/services/email.service.ts]: it would be better to send email from backend

**Review by tatianahutii — CHANGES_REQUESTED** [review]

File: backend/LeetWars.Emailer/Dockerfile (hunk 7 above)
Comment (SashaBondarchuk) [reply backend/LeetWars.Emailer/Dockerfile:1]: every our dockerfile has this, moreover I thought it will resolve sonarcloud security hotspot warning

File: frontend/src/app/core/services/email.service.ts (hunk 10 above)
Comment (SashaBondarchuk) [reply frontend/src/app/core/services/email.service.ts]: okay, but why this task include frontend part then?

Comment (tatianahutii) [reply frontend/src/app/core/services/email.service.ts]: because creating template required working with HTML and CSS =)

**Review by tatianahutii — APPROVED** [review]

**Review by Den313 — APPROVED** [review]
